src/chain/chain.rs

```diff
@@ -973,6 +980,351 @@ impl Chain {
+                    // should confirm if needed esp after msg_relay changes.
+                    section = section.into_iter().filter(is_connected).collect();
+                    return Ok(section);
+                }
+                candidates(target_name)
+            }
+            Authority::PrefixSection(ref prefix) => {
+                if prefix.is_compatible(&self.our_prefix()) {
+                    // only route the message when we have all the targets in our routing table -
+                    // this is to prevent spamming the network by sending messages with
+                    // intentionally short prefixes
+                    if !prefix.is_covered_by(self.prefixes().iter()) {
+                        return Err(Error::CannotRoute);
+                    }
+
+                    let is_compatible = |(ref pfx, section)| {
+                        if prefix.is_compatible(pfx) {
+                            Some(section)
+                        } else {
+                            None
+                        }
+                    };
+
+                    let mut targets = Iterator::flatten(
+                        self.all_sections()
+                            .filter_map(is_compatible)
+                            .map(SectionInfo::member_names),
+                    )
+                    .filter(is_connected)
+                    .collect::<BTreeSet<_>>();
+                    let _ = targets.remove(&self.our_id().name());
+                    return Ok(targets);
+                }
+                candidates(&prefix.lower_bound())
+            }
+        };
+        let target = self.get_routeth_node(&closest_section, dst.name(), Some(exclude), route)?;
+        Ok(iter::once(target).collect())
+    }
+
+    /// Returns our own section, including our own name.
+    pub fn our_section(&self) -> BTreeSet<XorName> {
+        self.opt_our_info()
+            .map_or_else(BTreeSet::new, SectionInfo::member_names)
+    }
+
+    /// Are we among the `count` closest nodes to `name`?
+    fn is_closest(&self, name: &XorName, count: usize, connected_peers: &[&XorName]) -> bool {
+        self.closest_names(name, count, connected_peers).is_some()
+    }
+
+    /// Returns whether we are a part of the given authority.
+    pub fn in_authority(&self, auth: &Authority<XorName>, connected_peers: &[&XorName]) -> bool {
+        match *auth {
+            // clients have no routing tables
+            Authority::Client { .. } => false,
+            Authority::ManagedNode(ref name) => self.our_id().name() == name,
+            Authority::ClientManager(ref name)
+            | Authority::NaeManager(ref name)
+            | Authority::NodeManager(ref name) => {
+                self.is_closest(name, self.min_sec_size, connected_peers)
+            }
+            Authority::Section(ref name) => self.our_prefix().matches(name),
+            Authority::PrefixSection(ref prefix) => self.our_prefix().is_compatible(prefix),
```

Comment:
I guess this is just ported, but should it be this way? Shouldn't we just check if the prefix matches our name? (IIRC `is_compatible` is symmetric, so if we are a node in section 0 with name starting with 00, this will still say that we are a part of PrefixSection(01), because 0 and 01 are compatible.)

Reply:
I'm not sure, I think this was done deliberately: The question is just what should happen if a message to prefix `01` reaches section `0`. I was thinking it should still be delivered to the whole section, but maybe you're right, and that's not necessary.

Reply:
Yeah, I'm now pretty sure it has been done deliberately and that it's actually a good idea.